--- coreutils-6.1/src/dircolors.hin
+++ coreutils-6.1/src/dircolors.hin
@@ -5,9 +5,9 @@
 # Free Software Foundation, Inc.
 # Copying and distribution of this file, with or without modification,
 # are permitted provided the copyright notice and this notice are preserved.
-
-# The keywords COLOR, OPTIONS, and EIGHTBIT (honored by the
-# slackware version of dircolors) are recognized but ignored.
+#
+# You can copy this file to .dir_colors in your $HOME directory to override
+# the system defaults.
 
 # Below, there should be one TERM entry for each termtype that is colorizable
 TERM Eterm
@@ -66,7 +66,8 @@
 DOOR 01;35	# door
 BLK 40;33;01	# block device driver
 CHR 40;33;01	# character device driver
-ORPHAN 40;31;01 # symlink to nonexistent file, or non-stat'able file
+ORPHAN 01;05;37;41  # orphaned syminks
+MISSING 01;05;37;41 # ... and the files they point to
 SETUID 37;41	# file that is setuid (u+s)
 SETGID 30;43	# file that is setgid (g+s)
 STICKY_OTHER_WRITABLE 30;42 # dir that is sticky and other-writable (+t,o+w)
@@ -86,6 +87,8 @@
 #.com 01;32
 #.btm 01;32
 #.bat 01;32
+#.sh  01;32
+#.csh 01;32
 
 .tar 01;31 # archives or compressed (bright red)
 .tgz 01;31
@@ -97,9 +100,18 @@
 .Z   01;31
 .gz  01;31
 .bz2 01;31
+.bz  01;31
+.tbz2 01;31
+.tz  01;31
 .deb 01;31
 .rpm 01;31
 .jar 01;31
+.rar 01;31  # rar
+.ace 01;31  # unace
+.zoo 01;31  # zoo
+.cpio 01;31 # cpio
+.7z  01;31  # p7zip
+.rz  01;31  # rzip
 
 # image formats
 .jpg 01;35
@@ -115,9 +127,26 @@
 .tif 01;35
 .tiff 01;35
 .png 01;35
+.mng 01;35
+.pcx 01;35
+.yuv 01;35
 .mov 01;35
 .mpg 01;35
 .mpeg 01;35
+.m2v 01;35  # MPEG-2 Video only
+.mkv 01;35  # Matroska (http://matroska.org/)
+.ogm 01;35  # Ogg Media File
+.mp4 01;35  # 'Offical' container for MPEG-4
+.m4v 01;35  # MPEG-4 Video only
+.mp4v 01;35 # MPEG-4 Video only
+.vob 01;35  # MPEG-2 DVD
+.qt 01;35   # Quicktime (http://developer.apple.com/qa/qtw/qtw99.html)
+.nuv 01;35
+.wmv 01;35  # Windows Media Video
+.asf 01;35  # Advanced Systems Format (contains Windows Media Video)
+.rm 01;35   # Real Media
+.rmvb 01;35 # Real Media Variable Bitrate
+.flc 01;35  # AutoDesk Animator
 .avi 01;35
 .fli 01;35
 .gl 01;35
@@ -125,9 +154,26 @@
 .xcf 01;35
 .xwd 01;35
 
+# Document files
+.pdf 00;32
+.ps 00;32
+.txt 00;32
+.patch 00;32
+.diff 00;32
+.log 00;32
+.tex 00;32
+.doc 00;32
+
 # audio formats
 .flac 01;35
 .mp3 01;35
-.mpc 01;35
-.ogg 01;35
-.wav 01;35
+.mpc 00;36
+.ogg 00;36
+.wav 00;36
+.mid 00;36
+.midi 00;36
+.au 00;36
+.flac 00;36
+.aac 00;36
+.ra 01;36
+.mka 01;36
