diff -urN 2/src/nice.c 3/src/nice.c
--- 2/src/nice.c	2005-03-25 23:51:05.000000000 +0000
+++ 3/src/nice.c	2005-03-30 14:30:59.000000000 +0000
@@ -114,6 +114,8 @@
       if (s[0] == '-' && ISDIGIT (s[1 + (s[1] == '-' || s[1] == '+')])
 	  && (posix2_version () < 200112 || !getenv ("POSIXLY_CORRECT")))
 	{
+	  if (!getenv ("POSIXLY_CORRECT"))
+		error (0, 0, _("`%s' option is obsolete; use `-n %s' since this will be removed in the future"), s, s+1);
 	  adjustment_given = s + 1;
 	  ++i;
 	}
