Bug #84564.  Fix df, du, ls not understanding BLOCK_SIZE=HUMAN as in the manpage.

--- coreutils-5.2.1/lib/human.c.az	2005-03-14 19:41:17.000000000 +0200
+++ coreutils-5.2.1/lib/human.c	2005-03-14 19:47:22.000000000 +0200
@@ -409,10 +409,11 @@
 # define DEFAULT_BLOCK_SIZE 1024
 #endif
 
-static char const *const block_size_args[] = { "human-readable", "si", 0 };
+static char const *const block_size_args[] = { "human-readable", "HUMAN-READABLE", "si", 0 };
 static int const block_size_opts[] =
   {
     human_autoscale + human_SI + human_base_1024,
+    human_autoscale + human_SI + human_base_1024,
     human_autoscale + human_SI
   };
 
